### ISSUE TITLE
Hash cache actually depends only on the MTime

### ISSUE THREAD

**Issue by A----** [getgrav/grav#2878]
Reading this ( https://learn.getgrav.org/16/advanced/performance-and-caching#grav-core-caching ), I was under the impression that the `hash` method would actually hash the files (in a fast not-so-secure way as it is not the point.)

I guess that the idea is to not load the file from disk, making it faster. Here comes the issue: when deploying, you have to be careful or preserving file attributes which isn't easy or borderline impossible (think checking out files from a SCM), if you don't want to invalidate an existing cache in those conditions. My use case: I build a docker image with the content. I set the mtime to a fixed date, so no invalidation (good), even when a file changes (bad).

So, without loading the file in a way or an other, we could add to the hash the filesize to the hash (which is present in the file stats) as a poor-man replacement.

Same should be done with the images.

Also, there might be a mistake on the `file` method documentation, it's written : “Grav uses a highly optimized routing to run through all the files” but it seems that `Folder::lastModifiedFile` is called without a second argument, filtering in only `md` and `yaml` files.

**Comment by mahagr** [comment]
Yes, it depends on the inode information, which is likely already cached by the operating system. If it would not do that, pages would all load super slow.

Even `file` cache is too slow on larger sites and should not be used. This is one the reasons why we're replacing the old page logic with new Flex pages. It will make even fewer checks relying on other methods of keeping track of the file updates.

**Comment by A----** [comment]
I tried the following patches:

```diff
--- system/src/Grav/Common/Filesystem/Folder.php
+++ system/src/Grav/Common/Filesystem/Folder.php
@@ -116,7 +116,7 @@
             $iterator = new \RecursiveIteratorIterator($directory, \RecursiveIteratorIterator::SELF_FIRST);
 
             foreach ($iterator as $file) {
-                $files[] = $file->getPathname() . '?'. $file->getMTime();
+                $files[] = [ $file->getPathname(), $file->getMTime(), $file->getSize() ];
             }
         }
```

```diff
--- system/src/Grav/Common/Page/Page.php
+++ system/src/Grav/Common/Page/Page.php
@@ -146,7 +146,7 @@
         $this->home_route = $this->adjustRouteCase($config->get('system.home.alias'));
         $this->filePath($file->getPathname());
         $this->modified($file->getMTime());
-        $this->id($this->modified() . md5($this->filePath()));
+        $this->id($file->getSize() . $this->modified() . md5($this->filePath()));
         $this->routable(true);
         $this->header();
         $this->date();
@@ -264,7 +264,7 @@
 
             // Reset header and content.
             $this->modified = time();
-            $this->id($this->modified() . md5($this->filePath()));
+            $this->id(strlen($var) . $this->modified() . md5($this->filePath()));
             $this->header = null;
             $this->content = null;
             $this->summary = null;
```

But switch to `check: off` and a clear-cache after deploy as I still has some issues with Twig `auto_reload`. It should hopefully work now, I might bring back the precache plugin now.
